pkg/controllers/config/configsync.go

```diff
@@ -0,0 +1,211 @@
+		return err
+	}
+
+	sourceConfig, err := c.configMapLister.ConfigMaps(util.OpenShiftConfigNamespace).Get(infra.Spec.CloudConfig.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: report error after some while?
+			klog.V(2).Infof("Waiting for config map %s from %s", infra.Spec.CloudConfig.Name, util.OpenShiftConfigNamespace)
+			return nil
+		}
+		return err
+	}
+
+	targetConfig, err := translateConfigMap(sourceConfig, isMultiAZDeployment)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = resourceapply.ApplyConfigMap(ctx, c.kubeClient.CoreV1(), c.eventRecorder, targetConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isMultiAZDeployment() (bool, error) {
+	ci, err := getCloudInfo()
+	if err != nil {
+		return false, fmt.Errorf("couldn't collect info about cloud availability zones: %w", err)
+	}
+	// We consider a cloud multiaz when it either have several different zones
+	// or compute and volumes are different.
+	return len(ci.ComputeZones) > 1 || len(ci.VolumeZones) > 1 || ci.ComputeZones[0] != ci.VolumeZones[0], nil
+}
+
+func translateConfigMap(cloudConfig *v1.ConfigMap, isMultiAZDeployment bool) (*v1.ConfigMap, error) {
+	content, ok := cloudConfig.Data[defaultConfigKey]
+	if !ok {
+		return nil, fmt.Errorf("OpenStack config map did not contain key %s", defaultConfigKey)
+	}
+
+	cfg, err := ini.Load([]byte(content))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read the cloud.conf: %w", err)
+	}
+
+	// Set the static, must-have keys in the '[Global]' section. If these are
+	// already set by the user then tough luck
+	global, _ := cfg.GetSection("Global")
+	if global != nil {
+		klog.Infof("[Global] section found; dropping any legacy settings...")
+		// Remove the legacy keys, once we ensure they're not overridden
+		for key, value := range map[string]string{
+			"secret-name":      "openstack-credentials",
+			"secret-namespace": "kube-system",
+			"kubeconfig-path":  "",
+		} {
+			if global.Key(key).String() != value {
+				return nil, fmt.Errorf("'[Global] %s' is set to a non-default value", key)
+			}
+			global.DeleteKey(key)
+		}
+	} else {
+		// This probably isn't common but at least handling this allows us to
+		// recover gracefully
+		global, err = cfg.NewSection("Global")
+		if err != nil {
+			return nil, fmt.Errorf("failed to modify the provided configuration: %w", err)
+		}
+	}
+	for key, value := range map[string]string{
+		"use-clouds":  "true",
+		"clouds-file": "/etc/openstack/secret/clouds.yaml",
+		"cloud":       "openstack",
+	} {
+		_, err = global.NewKey(key, value)
+		if err != nil {
+			return nil, fmt.Errorf("failed to modify the provided configuration: %w", err)
+		}
+	}
+
+	// Now, modify the '[BlockStorage]' section as necessary
+	blockStorage, _ := cfg.GetSection("BlockStorage")
+	if blockStorage != nil {
+		klog.Infof("[BlockStorage] section found; dropping any legacy settings...")
+		// Remove the legacy keys, once we ensure they're not overridden
```

Comment:
Looks like we're removing `trust-device-path` unconditionally. Should we make the operator degraded like we did with CCCMO when this option is set?